HACKING.md

```diff
@@ -0,0 +1,388 @@
+them to the clinic.  If no community members need help, you can also just watch
+tests written in real time.
+
+### Documentation
+
+Salt uses both docstrings, as well as normal reStructuredText files in the
+`salt/doc` folder for documentation. To build the docs and view them in your
+browser, simply run:
+
+    cd doc; make html; cd _build/html; python -m http.server
+
+The first time this will take a while because there are a *lot* of modules.
+Maybe you should go grab some dessert if you already finished that sandwich.
+But once Sphinx is done building the docs you can go to
+http://localhost:8000/contents.html and view the docs as they exist. If you
+make changes, you can simply run this:
+
+    cd -; make html; cd _build/html; python -m http.server
+
+And get your updated docs.
+
+If your change is a doc-only change, you can go ahead and commit/push your code
+and open a PR. Otherwise you'll want to write some tests and code.
+
+### Running Development Salt
+
+Note: If you run into any issues in this section, check the Troubleshooting
+section.
+
+If you're going to hack on the Salt codebase you're going to want to be able to
+run Salt locally. The first thing you need to do is install Salt as an editable
+pip install:
+
+    python -m pip install -e .
+
+This will let you make changes to Salt without having to re-install it.
+
+After all of the dependencies and Salt are installed, it's time to set up the
+config for development. Typically Salt runs as `root`, but you can specify
+which user to run as. To configure that, just copy the master and minion
+configs. We have .gitignore setup to ignore the `local/` directory, so we can
+put all of our personal files there.
+
+    mkdir -p local/etc/salt/
+
+Create a master config file as `local/etc/salt/master`:
+
+    cat <<EOF >local/etc/salt/master
+    user: $(whoami)
+    root_dir: $PWD/local/
+    publish_port: 55505
+    ret_port: 55506
+    EOF
+
+And a minion config as `local/etc/salt/minion`:
+
+    cat <<EOF >local/etc/salt/minion
+    user: $(whoami)
+    root_dir: $PWD/local/
+    master: localhost
+    id: saltdev
+    master_port: 55506
+    EOF
+
+Now you can start your Salt master and minion, specifying the config dir.
+
+    salt-master --config-dir=local/etc/salt/ --log-level=debug --daemon
+    salt-minion --config-dir=local/etc/salt/ --log-level=debug --daemon
+
+Now you should be able to accept the minion key:
+
+    salt-key -c local/etc/salt -Ay
+
+And check that your master/minion are communicating:
+
+    salt -c local/etc/salt \* test.version
+
+Rather than running `test.version` from your master, you can run it from the
+minion instead:
+
+    salt-call -c local/etc/salt test.version
+
+Note that you're running `salt-call` instead of `salt`, and you're not
+specifying the minion (`\*`), but if you're running the dev version then you
+still will need to pass in the config dir. Now that you've got Salt running,
+you can hack away on the Salt codebase!
+
```

Comment:
All of the above prep work could also be done as a nox target.... But, it's outside of the scope of these changes....

Reply:
Must be a copypasta error 😝

Reply:
Also opaque. I'd argue that it's better to teach them how to fish rather than give them an auto-fishing rod here :joy: 

(I mean, generally I'm in favor of automation, but... if you're going to be hacking on Salt, I want to make it easy, but also transparent)

Reply:
But yes, *far* outside the scope of these changes, lol. I wouldn't use nox for that anyway - my command line history is enough, and if I forget what the incantation is I can always come here and copy/paste again.